contracts/staking/contracts/src/staking_pools/MixinCumulativeRewards.sol

```diff
@@ -0,0 +1,292 @@
+    )
+        internal
+    {
+        if (isDependent) {
+            _addDependencyOnCumulativeReward(
+                poolId,
+                epoch,
+                mostRecentCumulativeRewardInfo
+            );
+        } else {
+            _removeDependencyOnCumulativeReward(
+                poolId,
+                epoch,
+                mostRecentCumulativeRewardInfo.cumulativeRewardEpoch
+            );
+        }
+    }
+
+    /// @dev Adds a dependency on a cumulative reward for a given epoch.
+    /// @param poolId Unique Id of pool.
+    /// @param epoch to remove dependency from.
+    /// @param mostRecentCumulativeRewardInfo Info on the most recent cumulative reward.
+    function _addDependencyOnCumulativeReward(
+        bytes32 poolId,
+        uint256 epoch,
+        IStructs.CumulativeRewardInfo memory mostRecentCumulativeRewardInfo
+    )
+        internal
+    {
+        // add dependency by increasing the reference counter
+        cumulativeRewardsByPoolReferenceCounter[poolId][epoch] = cumulativeRewardsByPoolReferenceCounter[poolId][epoch].safeAdd(1);
+
+        // set CR to most recent reward (if it is not already set)
+        _trySetCumulativeReward(
+            poolId,
+            epoch,
+            mostRecentCumulativeRewardInfo.cumulativeReward
+        );
+    }
+
+    /// @dev Removes a dependency on a cumulative reward for a given epoch.
+    /// @param poolId Unique Id of pool.
+    /// @param epoch to remove dependency from.
+    /// @param mostRecentCumulativeRewardEpoch Epoch of the most recent cumulative reward.
+    function _removeDependencyOnCumulativeReward(
+        bytes32 poolId,
+        uint256 epoch,
+        uint256 mostRecentCumulativeRewardEpoch
+    )
+        internal
+    {
+        // remove dependency by decreasing reference counter
+        uint256 newReferenceCounter = cumulativeRewardsByPoolReferenceCounter[poolId][epoch].safeSub(1);
+        cumulativeRewardsByPoolReferenceCounter[poolId][epoch] = newReferenceCounter;
+
+        // clear cumulative reward from state, if it is no longer needed
+        _tryUnsetCumulativeReward(poolId, epoch);
+    }
+
+    /// @dev Computes a member's reward over a given epoch interval.
+    /// @param poolId Uniqud Id of pool.
+    /// @param memberStakeOverInterval Stake delegated to pool by meber over the interval.
+    /// @param beginEpoch beginning of interval.
+    /// @param endEpoch end of interval.
+    /// @return rewards accumulated over interval [beginEpoch, endEpoch]
+    function _computeMemberRewardOverInterval(
+        bytes32 poolId,
+        uint256 memberStakeOverInterval,
+        uint256 beginEpoch,
+        uint256 endEpoch
+    )
+        internal
+        view
+        returns (uint256)
+    {
+        // sanity check
+        if (memberStakeOverInterval == 0) {
+            return 0;
+        }
+
+        // compute reward
+        IStructs.Fraction memory beginRatio = cumulativeRewardsByPool[poolId][beginEpoch];
```

Comment:
Q: This structure is sparse, right? Is it possible, through regular operation, to come into this function where `endEpoch` is null, but `beginEpoch` is not? That would seem like a gnarly result, since it would be a negative/division by zero? Should we have some logic to handle these cases?

Reply:
There shouldn't be a scenario where this happens under normal operations, but I will add more assertions in this function to future-proof this claim. At present, there would be a divide-by-zero exception raised by `SafeMath`.